doc/timekpr.8

.TH TIMEKPR 8 "NOVEMBER 2008" Linux "User Manuals"
.SH TIMEKPR 
timekpr \- keep control of computer usage
.SH SYNOPSIS
.B timekpr 
.SH DESCRIPTION
.B timekpr 
tracks and controls the computer usage of user accounts. 
Using 
.B timekpr-gui(8), 
the administrator can limit daily usage per user by configuring

.IP . 
a daytime range (e.g. from 8am to 7pm) when to user is allowed to work.

.IP . 
an elapsed time (e.g. 90 minutes) the user is allowed to work per day.

.PP 
.B timekpr
allows users to login or to stay logged in only as long as both limits are met.
As long as a user hasn't any limits configured,
.B timekpr
does not do any monitoring or other actions on that user.

To
.I disallow logins outside the defined daytime range,
.B timekpr
relies on the settings in the configuration file for the 
.I time
PAM-Module entered by 
.B timekpr-gui.

To disallow users 
.I to stay logged in over the defined daytime range,
or
.I to stay logged in over the defined elapsed time,
.B timekpr 
checks the process tree at regular intervals. When a certain 
user is logged in, the file
.I /var/lib/timekpr/<username>.time
is updated to contain the new
elapsed time. This time is then checked against the total allowed
(stored in 
.I /etc/timekpr/<username>
).


When
.B timekpr
discovers a user violating its limits, the user is warned and given a grace
period to save his work. After this grace period, the user is 
.I kicked out
by killing its processes. In addition, the file 
.I /var/lib/timekpr/<username>.logout
is touched to state the fact that the user has been logged out forcibly by 
.B timekpr.

When the user tries to login again (within its daytime range, but all elapsed time
used up), it is kicked out immediately and then locked by editing the
config file for the 
.I access
PAM-Module. Now the user cannot login any more.

.SH FILES
.I /etc/timekpr.conf
.RS 
The system wide configuration file. See
.BR timekpr.conf (5)
for further details.
.RE
.I /etc/security/time.conf
.I /etc/security/access.conf
.RS 
The system wide PAM Configuration files for the time and and access modules.

.SH DIAGNOSTICS
The logfile is quite verbose:
.I /var/log/timekpr.log
. The logfile can located elsewhere, check the config file.

.SH BUGS
.B timekpr 
is in an early development stage. Please check https://bugs.launchpad.net/timekpr/+bugs 


.SH AUTHORS
Chris Jackson wrote the original bash script.

Even Nedberg and Savvas Radevic wrote the current python implementation.

Nicolas Iselin wrote this manpage.

.SH SEE ALSO
.BR timekpr.conf (5),
.BR timekpr-gui (8),
The
.BR "Linux-PAM guide for system administrators" 
(modules 
.BR access
and
.BR limits)




debian/changelog
timekpr (0.3.0~ppa1~ubuntu1~beta4) jaunty; urgency=low

  * timekpr-client will no longer display two or more notifications at the same time
  * Added Finish translation

 -- Even Nedberg <even@undertow>  Sun, 12 Apr 2009 20:09:28 +0200

timekpr (0.3.0~ppa1~ubuntu1~beta3) jaunty; urgency=low

  * Added manpage (LP: #302770)

 -- Even Nedberg <even@undertow>  Sun, 12 Apr 2009 20:07:34 +0200

timekpr (0.3.0~ppa1~intrepid1~beta2) intrepid; urgency=low

  * Fixed (LP: #314061): hard reference to /etc/timekpr in timekpr-client, Properly this time
  * Added translation, now including Danish, German, French, Hungarian, Norwegian bookmal and Swedish

 -- Even Nedberg <even@undertow>  Fri, 10 Apr 2009 12:00:39 +0200

timekpr (0.3.0~ppa1~intrepid1~beta1) intrepid; urgency=low

  * Fixed Bug #330261: GUI fails to start if no normal users present
  * Added i18n support to timekpr-gui and timekpr-client, Bug #302782
  * Fixed bug #314061: hard reference to /etc/timekpr in timekpr-client
  * Fixed Bug #344538: timekpr limitations allow maximum 14h (Fixed by Savvas)
  * Build from revison #212

 -- Even Nedberg <even@lateralus>  Thu, 09 Apr 2009 17:42:00 +0200

timekpr (0.2.2~ppa1~intrepid1) intrepid; urgency=low

  * .deb file now reflects revision 186
  * timekpr-client gives less notifications, less annoying
  * Logrotate should work now, earlier versions installed config file to wrong location
  * Logrotate will compress old logfiles on second run

 -- Even Nedberg <even@lateralus>  Wed, 24 Dec 2008 11:49:02 +0100

timekpr (0.2.2~b7~ppa1~intrepid1) intrepid; urgency=low

  * Fixed a bug where timekpr would crash at midnight, (LP: #307839)
  * timekpr no longer truncates log file when it is restared. (LP: #302833)

 -- Even Nedberg <even@lateralus>  Sun, 14 Dec 2008 21:02:41 +0100

timekpr (0.2.2~b6~ppa1~intrepid1) intrepid; urgency=low

  * Unrestricted users no longer get notified, closes: Bug#306121
  * Logfile is kept over system reboot and timekpr restart, logfile is rotated. closes: Bug#302833
  * Timekpr-gui shows remaining time for a restricted user. closes: Bug#302774
  * Added option to remove time from a user, aka Penalty button. closes: Bug#302776

 -- Even Nedberg <even@lateralus>  Wed, 10 Dec 2008 20:25:24 +0100

timekpr (0.2.2~b5~ppa1~intrepid1) intrepid; urgency=low

  * Fixed typo in timekpr.py, def logOut 

 -- Even Nedberg <even@lateralus>  Fri, 05 Dec 2008 18:11:43 +0100

timekpr (0.2.2~b4~ppa1~intrepid1) intrepid; urgency=low

  * Fixed Bug#305055

 -- Even Nedberg <even@lateralus>  Fri, 05 Dec 2008 17:11:25 +0100

timekpr (0.2.2~b3) UNRELEASED; urgency=low

  * Rev #149
  * Includes timekpr-client
  * Fix for problems reported in Bug#287012 by WNino
  * Checks users even if /etc/timekpr/username does not exist

 -- Even Nedberg <code@nedberg.net>  Thu, 20 Nov 2008 20:51:18 +0100

timekpr (0.2.2~b2) UNRELEASED; urgency=low

  * Rev #121
  * Works in Ubuntu Hardy Heron 8.04.1 now (Bug #287012)
  * Removed debian/menu, using gksu package (Bug #286527)
  * First user with UserID == UID_MIN from login.defs is now hidden
    from the user list (Bug #286529)

 -- Savvas Radevic <vicedar@gmail.com>  Thu, 23 Oct 2008 20:51:18 +0200

timekpr (0.2.2~b1) UNRELEASED; urgency=low

  * Pushed version 0.2.2 (so I could set it as beta 1 release)
  * Testing deb package
  * Revision #111

 -- Savvas Radevic <vicedar@gmail.com>  Sun, 19 Oct 2008 16:25:16 +0200

timekpr (0.2.1) UNRELEASED; urgency=low

  * Testing deb package
  * Revision #110

 -- Savvas Radevic <vicedar@gmail.com>  Thu, 16 Oct 2008 17:49:58 +0200

timekpr (0.2.0) UNRELEASED; urgency=low

  * Testing deb package
  * Revision #108

 -- Savvas Radevic <vicedar@gmail.com>  Wed, 15 Oct 2008 21:58:04 +0200

timekpr (0.1.9) UNRELEASED; urgency=low

  * Testing deb package
  * Revision #106

 -- Savvas Radevic <vicedar@gmail.com>  Wed, 15 Oct 2008 19:36:01 +0200

timekpr (0.1.0ubuntu10) UNRELEASED; urgency=low

  * Testing pythonized package install/remove/purge

 -- Savvas Radevic <vicedar@gmail.com>  Sun, 28 Sep 2008 20:20:59 +0200

timekpr (0.1.0ubuntu9) UNRELEASED; urgency=low

  * Testing

 -- Savvas Radevic <vicedar@gmail.com>  Tue, 23 Sep 2008 17:28:46 +0200

timekpr (0.1.0ubuntu8) UNRELEASED; urgency=low

  * Revision #30:
  * More fixes to timekpr, kicking and locking after allowed hours also
    when username.allow does not exist
  * Fixed error in timekpr
  * Changed timekpr, restricted user will be kicked out faster when
    logging in after allowed time

 -- Savvas Radevic <vicedar@gmail.com>  Sun, 21 Sep 2008 17:39:31 +0200

timekpr (0.1.0ubuntu7) UNRELEASED; urgency=low

  * Old timekpr: timekpr.old (not in deb package)
  * Fixed timekpr, tested working
  * New script to unlock a user account manually: timekpr-unlock username
  * New logging function: logkpr 'message' - Log file: /var/log/timekpr.log
  * New lock/unlock functions: based on usermod -e (user expiration) - Would use
    -L to actually lock it, but it would just say 'wrong password'
  * Tidied up timekpr
  * New variable for lock function: lock_lasts='1 hour' 

 -- Savvas Radevic <vicedar@gmail.com>  Sat, 20 Sep 2008 14:11:56 +0200

timekpr (0.1.0ubuntu6) UNRELEASED; urgency=low

  * Removed exclamation marks which caused problems to timekpr
  * Cleaned up timekpr script

 -- Savvas Radevic <vicedar@gmail.com>  Fri, 19 Sep 2008 23:35:45 +0200

timekpr (0.1.0ubuntu5) UNRELEASED; urgency=low

  * Cleaned up timekpr and timekpr-gui
  * timekpr-gui shows users with userid within UID_MIN and UID_MAX
  * timekpr and timekpr-gui now check and create the default
    directory if necessary
  * Applied possible bugfix (LP: #272142)

 -- Savvas Radevic <vicedar@gmail.com>  Fri, 19 Sep 2008 19:35:41 +0200

timekpr (0.1.0ubuntu4) UNRELEASED; urgency=low

  * GUI: Fixed users list, removed sys and glob modules

 -- Savvas Radevic <vicedar@gmail.com>  Fri, 19 Sep 2008 13:35:03 +0200

timekpr (0.1.0ubuntu3) UNRELEASED; urgency=low


timekpr.py
#!/usr/bin/env python
""" The "daemon service" for timekpr.
    Copyright / License: See COPYRIGHT.txt
"""

import re
from time import strftime, sleep, localtime, mktime, time as timenow
from os.path import split as splitpath, isfile, isdir, getmtime
from os import popen, mkdir, kill, remove
from glob import glob
from threading import Timer

#If DEVACTIVE is true, it uses files from local directory
DEVACTIVE = False

#IMPORT
if DEVACTIVE:
    from sys import path
    path.append('.')
from timekprpam import * # timekprpam.py
from timekprcommon import * # timekprcommon.py

#timekpr.conf variables (dictionary variable)
VAR = getvariables(DEVACTIVE)

#Check if admin/root
checkifadmin()

#Check if it exists, if not, create it
if not isdir(VAR['TIMEKPRDIR']):
    mkdir(VAR['TIMEKPRDIR'])
if not isdir(VAR['TIMEKPRWORK']):
    mkdir(VAR['TIMEKPRWORK'])

# Lists keeping track of users who has been notified
notifiedusers = []

# Keep track of todays date
THISDAY = strftime("%Y%m%d")

def logkpr(string,clear = 0):
    #To log: logkpr("Something")
    #To clear file and log: logkpr("Something",1)
    if VAR['DEBUGME'] != 'True':
        return
    if clear == 1:
        l = open(VAR['LOGFILE'], 'w')
    else:
        l = open(VAR['LOGFILE'], 'a')
    nowtime = strftime('%Y-%m-%d %H:%M:%S ')
    l.write(nowtime + string +'\n')

def logOut(user, somefile = ''):
    logkpr('logOut called with user: %s and somefile: %s' % (user, somefile))
    if somefile != '':
        f = open(somefile, 'w').close()
    if issessionalive(user):
        logkpr('logOut: Attempting killing %s (SIGTERM)...' % user)
        #this is a pretty bad way of killing a users processes, but we warned 'em
        getcmdoutput('pkill -SIGTERM -u %s' % user)
        sleep(5)
        if issessionalive(user):
            logkpr('logOut: Process still there, attempting force-killing %s (SIGKILL)...' % user)
            getcmdoutput('pkill -SIGKILL -u %s' % user)


## Using Linux-PAM to lock and disable users
def getlocklasts():
    #Returns the VAR['LOCKLASTS'] variable in seconds
    t=re.compile('(\d+) (second|minute|hour|day|week|month|year)s?').match(VAR['LOCKLASTS'])
    if not t:
        exit('Error: locklasts value "%s" is badly formatted, should be something like "1 week" or "2 hours"' % VAR['LOCKLASTS'])
    #n = time length
    #m = second|minute|hour|day|week|month|year
    (n,m)=(int(t.group(1)),t.group(2))
    #variable dictionary: multiply
    multiply = {
        'second': n,
        'minute': n * 60,
        'hour': n * 60 * 60,
        'day': n * 60 * 60 * 24,
        'week': n * 60 * 60 * 24 * 7,
        'month': n * 60 * 60 * 24 * 30
    }
    #Return in seconds (integer)
    return multiply[m]

def lockacct(u):
    #Locks user and sets the date in a file
    logkpr('lockacct called for user %s' % u)
    lockfile = VAR['TIMEKPRDIR'] + '/' + u + '.lock'
    f = open(lockfile, 'w')
    f.close()
    lockuser(u) # timekprpam.py

def checklockacct():
    #Check if user should be unlocked and unlock them
    logkpr('checklockacct called')
    #Find *.lock in VAR['TIMEKPRDIR']
    s = VAR['TIMEKPRDIR'] + '/*.lock'
    l = glob(s)
    for f in l:
        #Get username from filename - os.path.split
        u = splitpath(f)[1].replace('.lock', '')
        lastmodified = getmtime(f) #Get last modified time from username.lock file
        #Get time when lock should be lifted
        dtlock = float(lastmodified + getlocklasts())
        dtnow = float(timenow())
        #If time now is great than or equal to the time when lock should be lifted
        if dtnow >= dtlock:
            logkpr('checklockacct: %s should be unlocked, unlocking..' % u)
            unlockuser(u)
            logkpr('checklockacct: removing %s.lock file..' % u)
            remove(f)

## File defs
def fileisok(fname):
    #File exists and is today's?
    if isfile(fname) and fromtoday(fname):
        return True
    return False

def getusers():
    u = getcmdoutput('users')
    u = u.split()
    u = set(u)
    return list(u)

def issessionalive(user):
    # Checking if session process still running
    # Should it check against username and pid?
    # Returns:    True if process is still there (user logged in),
    #        False if user has logged out
    for u in getusers():
        if u == user:
            return True
    return False

def gettime(tfile, username):
    #Adds time to the timefile
    if fileisok(tfile):
        logkpr("This day\'s %s.time file exists, adding time" % username)
        t = open(tfile)
        newtime = int(t.readline()) + VAR['POLLTIME']
    else:
        t = open(tfile, 'w')
        newtime = VAR['POLLTIME']
    t.close()
    writetime(tfile, newtime)
    return newtime

def writetime(tfile, time):
    #Write time to timefile
    f = open(tfile, 'w')
    f.write(str(time))

def threadit(sleeptime, command, *args):
    t = Timer(sleeptime, command, args)
    t.start()

def addnotified(u):
    #Adds username to notifiedusers list, so it does not re-notify them
    try:
        notifiedusers.index(u)
    except ValueError:
        notifiedusers.append(u)

def isnotified(u):
    #Checks if username is already in notifiedusers list
    try:
        notifiedusers.index(u)
    except ValueError:
        return False
    return True

def removenotified(u):
    #Removes username from notifiedusers list, so it does not re-notify them
    try:
        notifiedusers.index(u)
    except ValueError:
        return
    notifiedusers.remove(u)

logkpr('Starting timekpr version %s' % getversion())
logkpr('Variables: GRACEPERIOD: %s POLLTIME: %s DEBUGME: %s LOCKLASTS: %s' % (\
        VAR['GRACEPERIOD'],
        VAR['POLLTIME'],
        VAR['DEBUGME'],
        VAR['LOCKLASTS']))
logkpr('Directories: LOGFILE: %s TIMEKPRDIR: %s TIMEKPRWORK: %s TIMEKPRSHARED: %s' % (\
        VAR['LOGFILE'],
        VAR['TIMEKPRDIR'],
        VAR['TIMEKPRWORK'],
        VAR['TIMEKPRSHARED']))

while (True):
    # Check if any accounts should be unlocked and re-activate them
    checklockacct()
    # Check if we have passed midnight, ie new day
    if THISDAY != strftime("%Y%m%d"):
        THISDAY = strftime("%Y%m%d")

    # Get the usernames and PIDs of sessions
    for username in getusers():
        conffile = VAR['TIMEKPRDIR'] + '/' + username
        # Check if user configfile exists and if user was not already notified
        if not isnotified(username):
            logkpr('configuration file for %s exists' % username)
            # Read lists: from, to and limit
            limits, bfrom, bto = readusersettings(username, conffile)
            timefile = VAR['TIMEKPRWORK'] + '/' + username + '.time'
            allowfile = VAR['TIMEKPRWORK'] + '/' + username + '.allow'
            latefile = VAR['TIMEKPRWORK'] + '/' + username + '.late'
            logoutfile = VAR['TIMEKPRWORK'] + '/' + username + '.logout'

            time = int(gettime(timefile, username))
            '''Is the user allowed to be logged in at this time?
            We take it for granted that if they are allowed to login all day ($default_limit) then
            they can login whenever they want, ie they are normal users'''

            # Get current day index and hour of day
            index = int(strftime("%w"))
            hour = int(strftime("%H"))

            logkpr('User: %s Day-Index: %s Seconds-passed: %s' % (username, index, time))

            # Compare: is current hour less than the one in bfrom list?
            if (hour < bfrom[index]):
                logkpr('Current hour less than the defined hour in conffile for user %s' % username)
                if isfile(allowfile):
                    if not fromtoday(allowfile):
                        logkpr('Extended login hours detected from %s.allow, but not from today' % username)
                        threadit(0.5, logOut, username)
                        remove(allowfile)
                else:
                    # User has not been given extended login hours
                    logkpr('Extended hours not detected, %s not in allowed period from-to' %username)
                    threadit(0.5, logOut, username)

            # Compare: is current hour greater/equal to $to array?
            if (hour >= bto[index]):
                logkpr('Current hour greater than the defined hour in conffile for user %s' % username)
                # Has the user been given extended login hours?
                if isfile(allowfile):
                    if not fromtoday(allowfile):
                        logkpr('Extended login hours detected from %s.allow, but not from today' % username)
                        # Has the user been late-kicked today?
                        if isfile(latefile):
                            if fromtoday(latefile):
                                logkpr('User %s has been late-kicked today' % username)
                                threadit(0.5, logOut, username)
                                remove(allowfile)
                                #Lock account
                                lockacct(username)
                            else:
                                logkpr('User %s has NOT been late-kicked today' % username)
                                threadit(float(VAR['GRACEPERIOD']), logOut, username, latefile)
                                threadit(float(VAR['GRACEPERIOD']), remove, allowfile)
                                addnotified(username)
                                threadit(VAR['GRACEPERIOD'], removenotified, username)
                                lockacct(username)
                    else:
                        logkpr('Extended login hours detected - %s.allow is from today' % username)
                else:
                    # User has not been given extended login hours
                    logkpr('Extended hours and %s.allow file not detected, %s not in allowed period from-to' % (username, username))
                    if isfile(latefile) and fromtoday(latefile):
                        logkpr('User %s has been late-kicked today' % username)
                        threadit(0.5, logOut, username)
                        #Lock account
                        lockacct(username)
                    else:
                        logkpr('User %s has NOT been late-kicked today' % username)
                        threadit(float(VAR['GRACEPERIOD']), logOut, username, latefile)
                        addnotified(username)
                        threadit(VAR['GRACEPERIOD'], removenotified, username)
                        lockacct(username)

            # Is the limit exeeded
            # Also includes a fix for a bug that caused an unrestricted user to be kicked and locked at midnight
            if (time > limits[index] and not limits[index] == 86400):
                logkpr('Exceeded today\'s access login duration user %s' % username)
                # Has the user already been kicked out?
                if isfile(logoutfile):
                    logkpr('Found %s.logout' % username)
                    # Was he kicked out today?
                    if fromtoday(logoutfile):
                        logkpr('%s has been kicked out today' % username)
                        threadit(0.5, logOut, username)
                        #Lock account
                        lockacct(username)
                    else:
                        # The user has not been kicked out today
                        logkpr('%s has been kicked out, but not today' % username)
                        threadit(float(VAR['GRACEPERIOD']), logOut, username, logoutfile)
                        addnotified(username)
                        threadit(VAR['GRACEPERIOD'], removenotified, username)
                        lockacct(username)
                else:
                    # The user has not been kicked out before
                    logkpr('Not found: %s.logout' % username)
                    threadit(float(VAR['GRACEPERIOD']), logOut, username, logoutfile)
                    addnotified(username)
                    threadit(VAR['GRACEPERIOD'], removenotified, username)
                    lockacct(username)

    # Done checking all users, sleeping
    logkpr('Finished checking all users, sleeping for %s seconds' % VAR['POLLTIME'])
    sleep(VAR['POLLTIME'])

